cmd/snap-confine/seccomp-support.c

```diff
@@ -775,126 +107,12 @@ scmp_filter_ctx sc_prepare_seccomp_context(const char *filter_profile)
-
-	char profile_path[512];	// arbitrary path name limit
-	sc_must_snprintf(profile_path, sizeof(profile_path), "%s/%s",
-			 filter_profile_dir, filter_profile);
-
-	f = fopen(profile_path, "r");
-	if (f == NULL) {
-		fprintf(stderr, "Can not open %s (%s)\n", profile_path,
-			strerror(errno));
-		die("aborting");
-	}
-	// Note, preprocess_filter() die()s on error
-	preprocess_filter(f, &pre);
-
-	if (pre.unrestricted) {
-		seccomp_release(ctx);
-		ctx = NULL;
-		goto out;
-	}
-	// FIXME: right now complain mode is the equivalent to unrestricted.
-	// We'll want to change this once we seccomp logging is in order.
-	if (pre.complain) {
-		seccomp_release(ctx);
-		ctx = NULL;
-		goto out;
-	}
-
-	char buf[SC_MAX_LINE_LENGTH];
-	while (fgets(buf, sizeof(buf), f) != NULL) {
-		lineno++;
-
-		// skip policy-irrelevant lines
-		if (validate_and_trim_line(buf, sizeof(buf), lineno) == 0)
-			continue;
-
-		char *buf_copy = strdup(buf);
-		if (buf_copy == NULL)
-			die("Out of memory");
-
-		int pr_rc = parse_line(buf_copy, &sargs);
-		free(buf_copy);
-		if (pr_rc != PARSE_OK) {
-			// as this is a syscall whitelist an invalid syscall
-			// is ok and the error can be ignored
-			if (pr_rc == PARSE_INVALID_SYSCALL)
-				continue;
-			die("could not parse line");
-		}
-
-		rc = seccomp_rule_add_exact_array(ctx, SCMP_ACT_ALLOW,
-						  sargs.syscall_nr,
-						  sargs.length, sargs.arg_cmp);
-		if (rc != 0) {
-			rc = seccomp_rule_add_array(ctx, SCMP_ACT_ALLOW,
-						    sargs.syscall_nr,
-						    sargs.length,
-						    sargs.arg_cmp);
-			if (rc != 0) {
-				fprintf(stderr,
-					"seccomp_rule_add_array failed with %i for '%s'\n",
-					rc, buf);
-				errno = 0;
-				die("aborting");
-			}
-		}
-	}
-
- out:
-	if (f != NULL) {
-		if (fclose(f) != 0)
-			die("could not close seccomp file");
-	}
-	sc_map_destroy();
-	return ctx;
-}
-
-void sc_load_seccomp_context(scmp_filter_ctx ctx)
-{
-	int rc;
-	uid_t real_uid, effective_uid, saved_uid;
-
-	// if sc_prepare_seccomp_context() sees @unrestricted or @complain it bails
-	// out early and destroys the context object. In that case we have nothing
-	// to do.
-	if (ctx == NULL) {
-		return;
-	}
-
-	if (getresuid(&real_uid, &effective_uid, &saved_uid) != 0)
-		die("could not find user IDs");
-
-	// If not root but can raise, then raise privileges to load seccomp
-	// policy since we don't have nnp
-	debug("raising privileges to load seccomp profile");
-	if (effective_uid != 0 && saved_uid == 0) {
-		if (seteuid(0) != 0)
-			die("seteuid failed");
-		if (geteuid() != 0)
-			die("raising privs before seccomp_load did not work");
-	}
-	// load it into the kernel
-	debug("loading seccomp profile into the kernel");
-	rc = seccomp_load(ctx);
-	if (rc != 0) {
-		fprintf(stderr, "seccomp_load failed with %i\n", rc);
-		die("aborting");
+	struct sock_fprog prog = {
+		.len = num_read / sizeof(struct sock_filter),
+		.filter = (struct sock_filter *)bpf,
+	};
+	if (prctl(PR_SET_SECCOMP, SECCOMP_MODE_FILTER, &prog)) {
```

Comment:
As discussed I added opcode based validation and removed it again because the consensus is that this kind of validation is too brittle. The check for /var/lib/snapd/seccomp/bpf all the way down to "/" is implemented (and uncovered a bug in the current snap-confine code).